### DISCUSSION TITLE
installGoogleBarcodeScannerModule is stuck

### DISCUSSION THREAD

**Discussion by Ly-sg** [capawesome-team/capacitor-mlkit#98] [Q&A]
Hello,
i tried the demo app for the Mlkit to to test the barCode scanner on an android emulator, however it tells me that i don't have the googleBarcodeScannerModule, but when i click on the app's install button, the progress is stuck at 5 indefinetly.
Is this normal ? how long does it take normally to complete the installation ?

**Comment by robingenz** [comment]
5 means the installation is failed (see [docs](https://capawesome.io/plugins/mlkit/barcode-scanning/#googlebarcodescannermoduleinstallstate)). 
Take a look at Logcat to check if a an error message is printed.

**Reply by Ly-sg** [reply]
Thanks for the clarification about the status value, here's the log output : 

```
D/TaplEvents( 1250): TIS / TouchInteractionService.onInputEvent: MotionEvent { action=ACTION_DOWN, actionButton=0, id[0]=0, x[0]=87.98828, y[0]=555.97656, toolType[0]=TOOL_TYPE_FINGER, buttonState=0, classification=NONE, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=780650, downTime=780650, deviceId=3, source=0x5002, displayId=0, eventId=458826953 }
D/TaplEvents( 1250): TIS / TouchInteractionService.onInputEvent: MotionEvent { action=ACTION_UP, actionButton=0, id[0]=0, x[0]=87.98828, y[0]=555.97656, toolType[0]=TOOL_TYPE_FINGER, buttonState=0, classification=NONE, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=780714, downTime=780650, deviceId=3, source=0x5002, displayId=0, eventId=711916550 }
V/Capacitor/Plugin( 2706): To native (Capacitor plugin): callbackId: 39044422, pluginId: BarcodeScanner, methodName: installGoogleBarcodeScannerModule
V/Capacitor( 2706): callback: 39044422, pluginId: BarcodeScanner, methodName: installGoogleBarcodeScannerModule, methodData: {}
W/ActivityManager(  622): Unable to start service Intent { act=com.google.android.finsky.BIND_PLAY_MODULE_SERVICE pkg=com.android.vending } U=0: not found
E/ZappPhoneskyConn( 1739): Unable to bind to Phonesky
I/GmsDebugLogger( 1739): [76] 
W/ActivityManager(  622): Unable to start service Intent { act=com.google.android.finsky.BIND_PLAY_MODULE_SERVICE pkg=com.android.vending } U=0: not found
E/ZappPhoneskyConn( 1739): Unable to bind to Phonesky
E/DynamicModuleDownloader( 1739): Zapp module request failed: null
D/EGL_emulation( 2706): app_time_stats: avg=7611.28ms min=1.49ms max=251063.73ms count=33
I/Capacitor/Console( 2706): File: http://localhost/ - Line 327 - Msg: undefined
I/ChimeraConfigurator( 1739): Starting update, reason: 4 urgentFeatures: mlkit.barcode.ui:1
I/GmsDebugLogger( 1739): [73] 1801
W/ZappDownloader( 1739): No successful module downloads for requested features [mlkit.barcode.ui].
I/GmsDebugLogger( 1739): [30] [TfliteDynamiteDynamite.integ:231810005100800],[MlkitBarcodeUi.optional:225014100000],[VisionBarcodeDynamite.integ:221180404100800]
W/ChimeraConfigService( 1739): Retry attempt was throttled. [CONTEXT service_id=264 ]
I/GmsDebugLogger( 1739): [83] 1 [mlkit.barcode.ui]
V/Capacitor/BarcodeScannerPlugin( 2706): Notifying listeners for event googleBarcodeScannerModuleInstallProgress
I/Capacitor/Console( 2706): File: http://localhost/src_app_modules_barcode-scanning_barcode-scanning_module_ts.js - Line 304 - Msg: googleBarcodeScannerModuleInstallProgress [object Object]
D/EGL_emulation(  859): app_time_stats: avg=17924.67ms min=17924.67ms max=17924.67ms count=1
D/EGL_emulation(  859): app_time_stats: avg=2071.91ms min=2071.91ms max=2071.91ms count=1
```

**Reply by robingenz** [reply]
Does your device have Google Play Services?

**Reply by Ly-sg** [reply]
Not sure but i don't think it has since i can't access to playStore, the emulator seems to be really minimalistic android 33 device.
Is there a CLI command i can use to install the missing packages ?

**Reply by robingenz** [reply]
Okay, so that's the reason for the error. Google Play Services are needed. This SO thread should help you: https://stackoverflow.com/questions/14536595/how-to-download-google-play-services-in-an-android-emulator

**Reply by Ly-sg** [reply]
I got it to work after creating an AVD with a system image that has integrated playstore functionality
thank you very much for the quick responses !

**Reply by JLoveI** [reply]
Don't have google access in china

**Reply by robingenz** [reply]
@JLoveI Then simply do not use the Google Barcode Scanner module. Use the [startScan(...)](https://capawesome.io/plugins/mlkit/barcode-scanning/#startscan) method instead.

**Comment by AuxSistemasLitoral** [comment]
Si la app que estoy desarrollando no la subo a playstore (la instalo como app de terceros en un dominio propio ). que plugin puedo utilizar para leer códigos de barra y/o qr?
dado que el ejemplo que amablemente dejas me arroja siempres esto 
Uncaught (in promise): Error: The Google Barcode Scanner Module is not available. You must install it first.
Error: The Google Barcode Scanner Module is not available. You must install it first.
